### ISSUE TITLE
Sleep key can't be ignored

### ISSUE THREAD

**Issue by jolmg** [systemd/systemd#22936]
**systemd version the issue has been seen with**

> 250.4

**Used distribution**
 
> Archlinux

**Linux kernel version used** (`uname -a`)

 > Linux aspero 5.16.16-arch1-1 #1 SMP PREEMPT Mon, 21 Mar 2022 22:59:40 +0000 x86_64 GNU/Linux

**CPU architecture issue was seen on**
 > x86-64

**Expected behaviour you didn't see**
 > To not suspend when `HandleSuspendKey=ignore` and `systemd-logind[16899]: Suspend key pressed short.`

**Unexpected behaviour you saw**
 > It suspended when short pressing the suspend key, despite having `HandleSuspendKey=ignore` in /etc/systemd/logind.conf

**Steps to reproduce the problem**
 > 1. Set HandleSuspendKey=ignore in /etc/systemd/logind.conf
 > 2. reboot
 > 3. short-press suspend key

**Additional program output to the terminal or log subsystem illustrating the issue**

Here is the journal output of systemd with log level "debug" from the moment logind reports the sleep button press, to when the suspension starts:

https://gist.github.com/jolmg/73dbd0a0d706434bff9fd682695c26e3

Additionally,

```
$ systemd-analyze cat-config systemd/logind.conf
# /etc/systemd/logind.conf
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it under the
#  terms of the GNU Lesser General Public License as published by the Free
#  Software Foundation; either version 2.1 of the License, or (at your option)
#  any later version.
#
# Entries in this file show the compile time defaults. Local configuration
# should be created by either modifying this file, or by creating "drop-ins" in
# the logind.conf.d/ subdirectory. The latter is generally recommended.
# Defaults can be restored by simply deleting this file and all drop-ins.
#
# Use 'systemd-analyze cat-config systemd/logind.conf' to display the full config.
#
# See logind.conf(5) for details.

[Login]
#NAutoVTs=6
#ReserveVT=6
#KillUserProcesses=no
#KillOnlyUsers=
#KillExcludeUsers=root
#InhibitDelayMaxSec=5
#UserStopDelaySec=10
HandlePowerKey=ignore
HandleSuspendKey=ignore
HandleSuspendKeyLongPress=ignore
#HandleHibernateKey=hibernate
HandleLidSwitch=ignore
HandleLidSwitchExternalPower=ignore
HandleLidSwitchDocked=ignore
#HandleRebootKey=reboot
#HandleRebootKeyLongPress=poweroff
#PowerKeyIgnoreInhibited=no
#SuspendKeyIgnoreInhibited=no
#HibernateKeyIgnoreInhibited=no
#LidSwitchIgnoreInhibited=yes
#RebootKeyIgnoreInhibited=no
#HoldoffTimeoutSec=30s
#IdleAction=ignore
#IdleActionSec=30min
#RuntimeDirectorySize=10%
#RuntimeDirectoryInodesMax=400k
#RemoveIPC=yes
#InhibitorsMax=8192
#SessionsMax=8192
```

For context, I should mention that in my hardware, it's not actually a switch/button, but rather my laptop lid being opened. It's a convertible where the lid rotates sideways then closes so both the keyboard and display are facing upwards when converted. It's only when I open the lid while the display is facing upwards that the sleep key event is raised. It appears in `/dev/input/event*` as:

```
$ sudo evtest --grab /dev/input/event12
Input driver version is 1.0.1
Input device ID: bus 0x19 vendor 0x1 product 0x1 version 0x100
Input device name: "Panasonic Laptop Support"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 113 (KEY_MUTE)
    Event code 114 (KEY_VOLUMEDOWN)
    Event code 115 (KEY_VOLUMEUP)
    Event code 142 (KEY_SLEEP)
    Event code 148 (KEY_PROG1)
    Event code 205 (KEY_SUSPEND)
    Event code 224 (KEY_BRIGHTNESSDOWN)
    Event code 225 (KEY_BRIGHTNESSUP)
    Event code 236 (KEY_BATTERY)
    Event code 240 (KEY_UNKNOWN)
    Event code 431 (KEY_DISPLAYTOGGLE)
  Event type 4 (EV_MSC)
    Event code 4 (MSC_SCAN)
Properties:
Testing ... (interrupt to exit)
Event: time 1648806007.171376, type 4 (EV_MSC), code 4 (MSC_SCAN), value 07
Event: time 1648806007.171376, type 1 (EV_KEY), code 142 (KEY_SLEEP), value 1
Event: time 1648806007.171376, -------------- SYN_REPORT ------------
Event: time 1648806007.171390, type 4 (EV_MSC), code 4 (MSC_SCAN), value 07
Event: time 1648806007.171390, type 1 (EV_KEY), code 142 (KEY_SLEEP), value 0
Event: time 1648806007.171390, -------------- SYN_REPORT ------------
```

`systemd-inhibit --what=handle-suspend-key cat` also doesn't seem to stop it from suspending.

**Comment by jolmg** [comment]
The only things that have worked to prevent the sleep is grabbing the event device with the `evtest` command above and disabling the lid switch in the BIOS.

**Comment by jolmg** [comment]
Another thing that "worked" is that while having gdb attached to systemd-logind, if I took too long stepping through the code and not reaching wherever it is that the sleep is triggered, then the sleep wouldn't be triggered. IOW, if on a breakpoint in `button_dispatch()` I did `n` then held Enter, it would sleep in 3-5 seconds. However, if I took my time looking around, eventually when I do `c`, the program receives a SIGABRT and I have to restart gdb, but it doesn't sleep in this case.

**Comment by poettering** [comment]
what does "loginctl show-seat" say about the property?

**Comment by jolmg** [comment]
```
$ sudo loginctl show-seat
EnableWallMessages=no
NAutoVTs=6
KillUserProcesses=no
RebootToFirmwareSetup=no
RebootToBootLoaderMenu=18446744073709551615
IdleHint=yes
IdleSinceHint=0
IdleSinceHintMonotonic=0
InhibitDelayMaxUSec=5s
UserStopDelayUSec=10s
HandlePowerKey=ignore
HandleSuspendKey=ignore
HandleHibernateKey=hibernate
HandleLidSwitch=suspend
HandleLidSwitchExternalPower=suspend
HandleLidSwitchDocked=suspend
HoldoffTimeoutUSec=30s
IdleAction=ignore
IdleActionUSec=30min
PreparingForShutdown=no
PreparingForSleep=no
Docked=yes
LidClosed=no
OnExternalPower=yes
RemoveIPC=yes
RuntimeDirectorySize=1666293760
RuntimeDirectoryInodesMax=406810
InhibitorsMax=8192
NCurrentInhibitors=0
SessionsMax=8192
NCurrentSessions=1
```

**Comment by jolmg** [comment]
From what I could tell while stepping around with gdb last night, `manager_handle_action()` seemed to properly ignore the keypress. The laptop slept if from then on I did `n` then held Enter for a few seconds, That is, some event processing, many iterations later in `manager_run()`, caused the sleep.

Yesterday, I thought I was seeing the sleep being triggered by the call to `sd_event_get_state()`, because that's what was on the prompt when I re-woke the laptop. It didn't make any sense, but now that I'm rested I realize the prompt was prepared to run that and what had actually just ran was `sd_event_run()`, which makes a lot more sense.

I'll try identifying the event some time later today or this weekend.

**Comment by poettering** [comment]
HandleSuspendKeyLongPress=ignore is what you need to set.

This is actually documented, if you look in the man page. But we forgot to add the field to the default logind.conf. Will post PR adding that shortly.

**Comment by poettering** [comment]
→ #22975 

Closing, since codewise things actually work as they should...

**Comment by jolmg** [comment]
> HandleSuspendKeyLongPress=ignore is what you need to set.

It doesn't appear in the `loginctl show-seat` output, but if you look at the `systemd-analyze cat-config systemd/logind.conf` output I showed on the first post, I do have it set to ignore. In fact, before I added that, I would accidently trigger it by staying too long in a breakpoint before continuing. That caused hibernations. HandleSuspendKeyLongPress's default action is hibernation, not sleep. What I'm experiencing is the system going to sleep.

Here's what little I found from trying to identify the events:

```
$ sudo gdb -p $(pgrep systemd-logind) -x =(cat << EOF
b login/logind-button.c:220
commands
p ev.type
p ev.value
p ev.code
end
b libsystemd/sd-event/sd-event.c:3595
commands
p s->type
p *s->io.callback
end
b libsystemd/sd-bus/sd-bus.c:3612
EOF
)
GNU gdb (GDB) 11.2
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Attaching to process 627
Reading symbols from target:/usr/lib/systemd/systemd-logind...
Reading symbols from target:/usr/lib/systemd/libsystemd-shared-250.so...
Reading symbols from target:/usr/lib/libc.so.6...
(No debugging symbols found in target:/usr/lib/libc.so.6)
Reading symbols from target:/lib64/ld-linux-x86-64.so.2...
(No debugging symbols found in target:/lib64/ld-linux-x86-64.so.2)
Reading symbols from target:/usr/lib/libacl.so.1...
(No debugging symbols found in target:/usr/lib/libacl.so.1)
Reading symbols from target:/usr/lib/libblkid.so.1...
(No debugging symbols found in target:/usr/lib/libblkid.so.1)
Reading symbols from target:/usr/lib/libcap.so.2...
(No debugging symbols found in target:/usr/lib/libcap.so.2)
Reading symbols from target:/usr/lib/libcrypt.so.2...
(No debugging symbols found in target:/usr/lib/libcrypt.so.2)
Reading symbols from target:/usr/lib/libgcrypt.so.20...
(No debugging symbols found in target:/usr/lib/libgcrypt.so.20)
Reading symbols from target:/usr/lib/libip4tc.so.2...
(No debugging symbols found in target:/usr/lib/libip4tc.so.2)
Reading symbols from target:/usr/lib/libkmod.so.2...
(No debugging symbols found in target:/usr/lib/libkmod.so.2)
Reading symbols from target:/usr/lib/liblz4.so.1...
(No debugging symbols found in target:/usr/lib/liblz4.so.1)
Reading symbols from target:/usr/lib/libmount.so.1...
(No debugging symbols found in target:/usr/lib/libmount.so.1)
Reading symbols from target:/usr/lib/libcrypto.so.1.1...
(No debugging symbols found in target:/usr/lib/libcrypto.so.1.1)
Reading symbols from target:/usr/lib/libp11-kit.so.0...
(No debugging symbols found in target:/usr/lib/libp11-kit.so.0)
Reading symbols from target:/usr/lib/libpam.so.0...
(No debugging symbols found in target:/usr/lib/libpam.so.0)
Reading symbols from target:/usr/lib/libseccomp.so.2...
(No debugging symbols found in target:/usr/lib/libseccomp.so.2)
Reading symbols from target:/usr/lib/libzstd.so.1...
(No debugging symbols found in target:/usr/lib/libzstd.so.1)
Reading symbols from target:/usr/lib/liblzma.so.5...
(No debugging symbols found in target:/usr/lib/liblzma.so.5)
Reading symbols from target:/usr/lib/libm.so.6...
(No debugging symbols found in target:/usr/lib/libm.so.6)
Reading symbols from target:/usr/lib/libgpg-error.so.0...
(No debugging symbols found in target:/usr/lib/libgpg-error.so.0)
Reading symbols from target:/usr/lib/libz.so.1...
(No debugging symbols found in target:/usr/lib/libz.so.1)
Reading symbols from target:/usr/lib/libgcc_s.so.1...
Reading symbols from target:/usr/lib/libffi.so.8...
(No debugging symbols found in target:/usr/lib/libffi.so.8)
Reading symbols from target:/usr/lib/libdl.so.2...
(No debugging symbols found in target:/usr/lib/libdl.so.2)
Reading symbols from target:/usr/lib/libpthread.so.0...
(No debugging symbols found in target:/usr/lib/libpthread.so.0)
Reading symbols from target:/usr/lib/libaudit.so.1...
(No debugging symbols found in target:/usr/lib/libaudit.so.1)
Reading symbols from target:/usr/lib/libcap-ng.so.0...
(No debugging symbols found in target:/usr/lib/libcap-ng.so.0)
Reading symbols from target:/usr/lib/libnss_systemd.so.2...
(No debugging symbols found in target:/usr/lib/libnss_systemd.so.2)
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
0x00007f0de39ec9ae in epoll_wait () from target:/usr/lib/libc.so.6
Breakpoint 1 at 0x561aabc9e912: file ../systemd-stable/src/login/logind-button.c, line 220.
Breakpoint 2 at 0x7f0de3d7063f: file ../systemd-stable/src/libsystemd/sd-event/sd-event.c, line 3595.
Breakpoint 3 at 0x7f0de3d161d2: file ../systemd-stable/src/libsystemd/sd-bus/sd-bus.c, line 3612.
(gdb) c
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf73410) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
warning: Source file is more recent than executable.
3595	        switch (s->type) {
$1 = SOURCE_IO
$2 = {int (sd_event_source *, int, uint32_t, void *)} 0x561aabc9e8b0 <button_dispatch>
(gdb) c
Continuing.

Breakpoint 1, button_dispatch (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=0x561aacf72ea0) at ../systemd-stable/src/login/logind-button.c:220
warning: Source file is more recent than executable.
220	        if (ev.type == EV_KEY && ev.value > 0) {
$3 = 1
$4 = 1
$5 = 142
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf73410) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$6 = SOURCE_IO
$7 = {int (sd_event_source *, int, uint32_t, void *)} 0x561aabc9e8b0 <button_dispatch>
(gdb) 
Continuing.

Breakpoint 1, button_dispatch (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=0x561aacf72ea0) at ../systemd-stable/src/login/logind-button.c:220
220	        if (ev.type == EV_KEY && ev.value > 0) {
$8 = 0
$9 = 0
$10 = 0
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf5e520) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$11 = SOURCE_IO
$12 = {int (sd_event_source *, int, uint32_t, void *)} 0x7f0de3d16139 <io_callback>
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf73410) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$13 = SOURCE_IO
$14 = {int (sd_event_source *, int, uint32_t, void *)} 0x561aabc9e8b0 <button_dispatch>
(gdb) 
Continuing.

Breakpoint 1, button_dispatch (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=0x561aacf72ea0) at ../systemd-stable/src/login/logind-button.c:220
220	        if (ev.type == EV_KEY && ev.value > 0) {
$15 = 1
$16 = 0
$17 = 142
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf73410) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$18 = SOURCE_IO
$19 = {int (sd_event_source *, int, uint32_t, void *)} 0x561aabc9e8b0 <button_dispatch>
(gdb) 
Continuing.

Breakpoint 1, button_dispatch (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=0x561aacf72ea0) at ../systemd-stable/src/login/logind-button.c:220
220	        if (ev.type == EV_KEY && ev.value > 0) {
$20 = 0
$21 = 0
$22 = 0
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf5e520) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$23 = SOURCE_IO
$24 = {int (sd_event_source *, int, uint32_t, void *)} 0x7f0de3d16139 <io_callback>
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf5e520) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$25 = SOURCE_IO
$26 = {int (sd_event_source *, int, uint32_t, void *)} 0x7f0de3d16139 <io_callback>
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf5e520) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$27 = SOURCE_IO
$28 = {int (sd_event_source *, int, uint32_t, void *)} 0x7f0de3d16139 <io_callback>
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf5e150) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$29 = SOURCE_TIME_MONOTONIC
$30 = {int (sd_event_source *, int, uint32_t, void *)} 0x7f0de3d161d2 <time_callback>
(gdb) 
Continuing.

Breakpoint 3, time_callback (s=0x561aacf5e150, usec=0, userdata=0x561aacf53f80) at ../systemd-stable/src/libsystemd/sd-bus/sd-bus.c:3612
warning: Source file is more recent than executable.
3612	static int time_callback(sd_event_source *s, uint64_t usec, void *userdata) {
(gdb) 
Continuing.

Breakpoint 2, source_dispatch (s=s@entry=0x561aacf5e150) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3595
3595	        switch (s->type) {
$31 = SOURCE_TIME_MONOTONIC
$32 = {int (sd_event_source *, int, uint32_t, void *)} 0x7f0de3d161d2 <time_callback>
(gdb) 
```

It's at the first `SOURCE_TIME_MONOTONIC` that it goes to sleep.

I figure the SOURCE_IO events are part of it, and that they probably correspond to messages on dbus. I tried looking into that via wireshark, and I see this as the first message that happened on lifting the lid:

![imgcap-7P0JZOAK](https://user-images.githubusercontent.com/17855790/161905592-94674622-9d5d-44ae-b937-b8c39d01e446.png)

**Comment by jolmg** [comment]
~Pretty confused about this. The events that happen between the latter `/dev/input/event*` `EV_SYN` event (where the `ev` values I print are all 0) and the `SOURCE_TIME_MONOTONIC` that causes the sleep are very variable depending on how quickly I hit `Enter` to continue between them. If I hit `Enter` rapidly I get a lot more `SOURCE_IO`s and even some `SOURCE_TIME_MONOTONIC` that do nothing. Sometimes I'll also get a `SOURCE_DEFER` with a `async_polkit_defer` callback. If I step around and inspect too slowly, it can cancel out the sleep. This is probably why it feels ineffective to step through every event in detail. I can't do it at leisure because it cancels out the sleep and I don't know the relevancy of each event to what's going on. I imagine the `SOURCE_TIME_MONOTONIC` is some sort of timer to cause the sleep, but stepping through that is probably pointless, since what I need to know is what set it up.~

~Another thing that's confusing, when I start gdb, after the breakpoints have been set-up and I do the first "continue" without triggering the sleep with the lid, it just hangs for quite a long time and hits the breakpoints very rarely, meaning that there are hardly any irrelevant events going through the breakpoints. So, I would think that practically all events are relevant to the sleep. I guess the inconsistencies are because they're very time-dependent.~

I'm going to feel so dumb if this actually ends up being a configuration issue or due to some package I have. It's quite possible-----

Damn. Yes. I'm sorry, guys. Wow, I was right. I really do feel dumb. I have `bindsym XF86Sleep exec systemctl suspend` on my i3's config from back when I used a keyboard with a sleep key. I hope to one day outlive the embarrassment. Thanks for the help.

**Comment by poettering** [comment]
> It doesn't appear in the `loginctl show-seat` output,

This is fixed by #22975 actually.
